docs/DesignDocs/Attestation_API_Proposal.md

```diff
@@ -0,0 +1,590 @@
+  - This entry point has the same definition as in the OE SDK V0.9 release,
+  except that there are two returned claims lists: one for well-known base
+  claims, and the other for custom claims.
+    - Note: the separate lists ensures custom claims will not have name
+    conflict with base claims.
+- Verify a legacy-format report
+  - Entry point `oe_result_t (*verify_report)(...)`
+  - Note: this entry point must be implemented if the OS SDK framework needs to
+  support the legacy public API `oe_verify_report()` using the plugin.
+- Free a dynamically allocated claims list buffer.
+  - The same entry point definition as in the OE SDK V0.9 release.
+
+Use cases for a verifier plugin to interact with the OE SDK framework:
+- Register a verifier plugin.
+  - Helper function
+  `oe_result_t oe_register_verifier(oe_verifier_t* plugin, const void* config_data, size_t config_data_size)`.
+  - The same definition as in the OE SDK V0.9 release, except that this function
+  is not part of the public API.
+- Unregister a verifier plugin.
+  - Helper function
+  `oe_result_t oe_unregister_verifier(oe_verifier_t* plugin)`.
+  - The same definition as in the OE SDK V0.9 release, except that this function
+  is not part of the public API.
+
+In the OE SDK V0.10 release, initialization of verifier plugin libraries is
+triggered with the application call to `oe_verifier_initialize()`.
+The initialization function of every linked library enumerates its supported
+evidence formats and registers their plugins with the OE SDK framework via
+helper function `oe_register_verifier()`.
+
+# Implementation of SGX Plugins
+
+## Existing OE SDK V0.9 Implementation
+
+Only one attester plugin and one verifier plugin are implemented,
+for a single SGX ECDSA evidence format identified by `OE_SGX_PLUGIN_UUID`.
+
+The attester plugin entry points are implemented in
+[enclave/sgx/attester.c](https://github.com/openenclave/openenclave/blob/v0.9.x/enclave/sgx/attester.c).
+The plugin entry points are returned in an `oe_attester_t` structure by the
+implementation of the public API function `oe_sgx_plugin_attester()`. This
+attester plugin implementation is linked to the OE SDK enclave-side static
+library.
+
+The verifier plugin entry points are implemented in
+[common/sgx/verifier.c](https://github.com/openenclave/openenclave/blob/v0.9.x/common/sgx/verifier.c).
+The plugin entry points are returned in an `oe_verifier_t` structure by the
+implementation of the public API function `oe_sgx_plugin_verifier()`. This
+verifier plugin implementation is linked to both the OE SDK enclave-side and host-side
+static libraries.
+
+## Proposed V0.10 Implementation
```

Comment:
In order for encourage adoption, I suggest creating samples that shows how to use this API and/or have an implementation guide.

Reply:
The design document has a section "Application Developer Experience" on a possible application scenario, and ctests will be updated to call the proposed APIs.

Reply:
I'll note here that the design documents are not generally used as reference documents by end-users, nor are tests; the addition of samples is a requirement for public APIs per the [ApiGuidelines.md](https://github.com/openenclave/openenclave/blob/master/docs/ApiGuidelines.md) and are strongly recommended for experimental APIs. Particularly in this case, the guidance seems relevant:

>Having working samples may help others evaluate and determine how ready it is for prime time before the API moves from experimental to public.

Reply:
@CodeMonkeyLeet agreed! the proposed APIs will still be experimental, and the sample in https://github.com/openenclave/openenclave/tree/master/samples/remote_attestation will be adapted to use the new API. Maybe in this sample, we can have support of both the legacy (production) and new (experimental) APIs and use a macro to switch between them?

Reply:
I think it's probably easier for the user of the samples to have them as separate samples instead of having to deal with even more instructions to build one version instead of the other. If the expectation is that the new samples will be direct replacements for the old ones, then I think it's fine to create a duplicate of the old samples and just use the new API calls, although that creates a bit more maintenance overhead. I would also suggest refactoring out shared code in that case that might be reusable as shared library components between samples.

Reply:
@CodeMonkeyLeet agreed that we can have new sample to showcase the new API, and refactor out reusable shared code to minimize maintenance overhead. The samples will be addressed in @yentsanglee's issue #2975